content/blog/2022/2022-08-02-performance-with-completablefutures.md

```diff
@@ -0,0 +1,249 @@
+---
+title: "How to use CompletableFuture for better performance during API calls"
+categories: ["Java"]
+date: 2022-02-08T11:00:00 
+authors: [Bledi]
+excerpt: "In this article we will see ho to increase the performance of our application using the class
+CompletableFuture for remote API calls."
+url: api-calls-performance-with-completable-future
+---
+
+# How to use CompletableFuture for better performance during API calls
+
+## What is a Future ?
+
+The _Future_ is a Java interface that was introduced in Java 5 to represent some value that will be available in a
+second moment. The advantages of using it are enormous because we could do some very intensive computation
+asynchronously without blocking the Thread that in the meantime can do some other useful job. We can think of it as
+going to the restaurant. During the time that the chef is preparing our dinner, we can do other things, like talking to
+friends or drinking a glass of wine and when in a second moment the chef has finished the preparation, we can finally
+eat. Another advantage is that using the Future interface is much more friendly than working with Threads.
+
+## CompletableFuture vs. Future
+
+The Future interface comes with its limitations. Let's see the most important ones :
+
+- We cannot complete manually the operation by providing a value
+- We cannot combine one or more asynchronous operations
+- There is no way to react to the completion of the asynchronous operation without blocking the thread by
+  invoking [get](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#get())
+  method of the _Future_ interface
+
+All the above limitations are made possible by the new class _CompletableFuture_ that was introduced with  
+version 8 of Java. CompletableFuture implements both Future
+and _[CompletionStage](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html)_ interfaces.
+Using this class we can now create a pipeline of asynchronous operations. Let's see a simple example of how to do it :
+
+```java
+class Demo {
+    public static void main(String[] args) {
+
+        CompletableFuture.supplyAsync(() -> "hello completable future")
+                .thenApply(String::toUpperCase)
+                .thenAccept(System.out::println);
+    }
+
+}
+```
+
+In the example above we can notice how simple is to create such a pipeline. We are firstly calling the `supplyAsync`
+method which takes a `Supplier` and returns a new CompletableFuture, we are then transforming the result to uppercase
+String by calling `thenApply` method and in the end, we just print the value on the console using `thenAccept` that
+takes a Consumer as the argument. If we think for a moment, working with CompletableFuture is very similar to Java
+Streams.
+
+## Calling remote APIs without using CompletableFuture (Synchronously)
```

Comment:
Since this and the next 2 sections are talking about the same example application, I suggest to create a top-level section "Performance Gains when Using `CompletableFuture`" and make these three sections subsections of that one. 
i.e.

* Performance Gains With `CompletableFuture`
  * Performance Using Synchronous Calls
  * Performance Using a Parallel Stream
  * Performance Using Completable Future

Reply:
done